tests/lms/util/_lti_launch_test.py

```diff
@@ -0,0 +1,89 @@
+from unittest import mock
+
+import pytest
+from sqlalchemy.orm.exc import NoResultFound
+from pylti.common import LTIException
+
+from lms.util import lti_launch
+from lms.exceptions import MissingLTILaunchParamError
+from lms.models import ApplicationInstance
+
+
+class TestLTILaunch:
+    @pytest.mark.usefixtures("application_instance")
+    def test_it_raises_if_theres_no_oauth_consumer_key(self, pyramid_request, wrapper):
+        del pyramid_request.params["oauth_consumer_key"]
+
+        with pytest.raises(MissingLTILaunchParamError, match="oauth_consumer_key"):
+            wrapper(pyramid_request)
+
+    def test_it_crashes_if_theres_no_application_instance_in_the_db(
```

Comment:
I'm having trouble tracking down where in the decorator function this lookup and Exception occurs. It is the `shared_secret` call? https://github.com/hypothesis/lms/blob/8781815270b52514872a7a91e6bc1951c801f0c6/lms/util/lti_launch.py#L52

It's not immediately obvious from the naming of that that it has a DB dependency. Also, should this be mocked? Sorry if I've got this totally wrong...

Reply:
`get_secret()` calls `get_application_instance()` which queries the DB. `lms.util.get_application_instance()` is also a near-exact duplicate of another shared function in a completely different location that also gets `ApplicationInstance`'s from the DB: `lms.models.application_instance.find_by_oauth_consumer_key()` :)

But I think you're right it should be mocked...

Reply:
Done